### ISSUE TITLE
Work flow of smart-home-nodejs

### ISSUE THREAD

**Issue by Nguyen-Cao-Bai** [actions-on-google/smart-home-nodejs#489]
Dear all,

I just jointed to "Smart home project", and I am working with Nodejs environment.
"smart-home-nodejs" is very helpful for reference.
However, there many parts (Firebase, gg action, userID, token,....)
Can you show me the work flow of smart-home-nodejs (link-link)?
I think it is easy to understand!

Thank you so much!

**Comment by Fleker** [comment]
The [documentation](https://developers.google.com/assistant/smarthome/overview) may be a good place to start. The [codelab](https://codelabs.developers.google.com/codelabs/smarthome-washer/#0) also walks through each of the sections in a more step-by-step manner which may be easier to follow.

**Comment by Nguyen-Cao-Bai** [comment]
Good morning Fleker,

Thank for answer, I already seen them some days ago.
I am looking for the working flow/mechanism to know how all parts connect/link.

And one more, smart-home-nodejs is a sample project, I think it free.
In the fact, when we deploy a business project, which step/part need to be paid?

**Comment by Fleker** [comment]
Here's a quick summary of each part:

* Firebase - Used primarily as a database service.
* GA Action - Used as the interface with the Google Assistant
* userID (agentUserId) - The identifier you have for your device on your service (their email?)
* Token (auth token) - A unique token that tells your service which of your users is making the request

Generally the flow is:
* User visits the Google Assistant Action and links their account using their user ID
* User visits your OAuth page and signs in
* Your service returns a refresh token
* Google Assistant sends your service the refresh token
* Your service returns an access token
* Google Assistant sends a SYNC request, looking for the user's devices
* Your service gets the devices (from a database perhaps)

There is not a charge for deploying a project to production.

**Comment by Nguyen-Cao-Bai** [comment]
Good morning Fleker,

Sorry for reply lately,
Thank for your answer, let me take a look.
However, 
>> There is not a charge for deploying a project to production.
I think we will use Google service/cloud for production
and we no need to pay anything?

Someone said that we need to pay 5$/month for API key, 
sorry I am not sure :(

**Comment by Fleker** [comment]
I'm not sure who told you that. At least with respect to Google there aren't any charges necessarily. If you're using additional services they may have fees.

**Comment by Nguyen-Cao-Bai** [comment]
Good afternoon Fleker,

Sorry for reply lately,
When we add my project on Google assistance app (phone), we see that it begin with [test]
ex: project name "luck home" => it will show as "[test] lucky home"
I am not sure, I think we need to pay something to commercialize project, also remove [test] label.

One more in source code, in "auth-provider.ts" we will get authorization/accessToken from "header"
and this is imported from 'actions-on-google'. What is this "header", where can we find out it?

Thank for your help

**Comment by Fleker** [comment]
Right now your project is in a test state, as you're testing it. When you're ready for production, follow the [guide on testing and submission](https://developers.google.com/assistant/smarthome/develop/testing-submitting).

The `header` is the HTTP header in the request that your server receives.

**Comment by Nguyen-Cao-Bai** [comment]
Hi Flexker,

I just found the topic where I got information of google payment ($5/month).
Please refer to below link, sorry for Vietnamese:
https://lbminhautomation.com/lien-ket-google-assistant-va-hass.html

I think 5$/month is Google Home Assistant Cloud service.

And what is difference between "API key" and "Service account key" 
at Create Credentials on Google Cloud Console API & Services page?

Nice day!

**Comment by Fleker** [comment]
The $5/month seems to be specifically for the _Home Assistant_ service, which is a service provider (think Philips Hue or Roomba) and unrelated to the _Google Assistant_ platform.

The *API Key* is a single string for making authenticated requests. The service account key is more like a username/password for making authenticated requests. The latter is recommended as it is more secure.

**Comment by Nguyen-Cao-Bai** [comment]
Hi Flexker,

Good afternoon Fleker,

Sorry for reply lately, I just come back for this project.
I have one question when run your example project.
If we buy some Wifi devices from other makers (TUYA, XIAOMI, PHILLIPS,...) which compatible with Google Assistance and add them to Google home application (on mobile).
Of course, in this case I think we will see them (list of these devices) on Google home application (on mobile), there are two types of device:
 1. Simulation devices (you designed by code)
 2. Real devices as TUYA, XIAOMI, PHILLIPS,...

At that time, how on your Web-app (front-end)? 
CAN we see list of these devices, both of "1" and "2", or still just "1"?

_**In conclusion, CAN we sync list of devices from Google cloud to Developer cloud?**_
Purpose: we want to control NOT only us developed devices, BUT also 3rd devices by us developer Web-app!
3rd Devices <---> 3rd Cloud <---> Google cloud <---> Developer cloud

Thanks!

**Comment by Fleker** [comment]
You, as a developer, can only access devices that are part of your service. Devices from other companies are not accessible.

**Comment by Nguyen-Cao-Bai** [comment]
Dear Fleker,
I think other devices which compatible with Google Assistance will use a common library to communication (can be "on-gooogle-action", not sure). So we can clasify "3rd Devices <---> 3rd Cloud <---> Google cloud <---> Developer cloud" as
1. 3rd Devices <---> 3rd Cloud: 3rd library
2. 3rd Cloud <---> Google cloud: common library (support by google)
3. Google cloud <---> Developer cloud: same as 2
Because device is compatible with Google assistance, so it have to support a mechanism to convert a command from common library to 3rd library, and we don't care and can not accept this mechanism.
Developer job is only send command as common library format.
Sorry if my talking is not clear to get it :(

**Comment by Fleker** [comment]
Not every developer will want their devices to be controlled by other developers arbitrarily.

**Comment by Nguyen-Cao-Bai** [comment]
Dear Fleker,
As project which my team already done before, we use Zigbee connectivity with Zigbee 3.0 library. In this case, we can control every 3rd devices as long as they also use Zigbee 3.0. For wifi device, base on this thinking, have something wrong?

**Comment by Fleker** [comment]
Devices are controlled primarily through secure connections between Google and a developer cloud. Users can create their own routines to orchestrate multiple devices, but developers cannot arbitrarily control every device that belong to a user.

**Comment by Nguyen-Cao-Bai** [comment]
Dear Flexker,

Thank for your helpful answer.
Let me find out more about this matter.
We will discuss again when have something new.
 
Thank you so much!
